src/blas/backends/rocblas/rocblas_scope_handle.cpp

```diff
@@ -0,0 +1,154 @@
+            if (handle != nullptr) {
+                ROCBLAS_ERROR_FUNC(rocblas_destroy_handle, err, handle);
+                handle = nullptr;
+            }
+            else {
+                delete handle_pair.second;
+            }
+            handle_pair.second = nullptr;
+        }
+    }
+    rocblas_handle_container_mapper_.clear();
+}
+
+/**
+ * Inserts a new element in the map if its key is unique. This new element
+ * is constructed in place using args as the arguments for the construction
+ * of a value_type (which is an object of a pair type). The insertion only
+ * takes place if no other element in the container has a key equivalent to
+ * the one being emplaced (keys in a map container are unique).
+ */
+thread_local rocblas_handle_container<pi_context> RocblasScopedContextHandler::handle_helper =
+    rocblas_handle_container<pi_context>{};
+
+RocblasScopedContextHandler::RocblasScopedContextHandler(cl::sycl::queue queue,
+                                                         cl::sycl::interop_handle &ih)
+        : interop_h(ih),
+          needToRecover_(false) {
+    placedContext_ = new cl::sycl::context(queue.get_context());
+    auto device = queue.get_device();
+    auto desired = cl::sycl::get_native<cl::sycl::backend::hip>(*placedContext_);
+    hipError_t err;
+    HIP_ERROR_FUNC(hipCtxGetCurrent, err, &original_);
+    if (original_ != desired) {
+        // Sets the desired context as the active one for the thread
+        HIP_ERROR_FUNC(hipCtxSetCurrent, err, desired);
+        // No context is installed and the suggested context is primary
+        // This is the most common case. We can activate the context in the
+        // thread and leave it there until all the PI context referring to the
+        // same underlying rocblas primary context are destroyed. This emulates
+        // the behaviour of the rocblas runtime api, and avoids costly context
+        // switches. No action is required on this side of the if.
+        needToRecover_ = !(original_ == nullptr);
+    }
+}
+
+RocblasScopedContextHandler::~RocblasScopedContextHandler() noexcept(false) {
+    if (needToRecover_) {
+        hipError_t err;
+        HIP_ERROR_FUNC(hipCtxSetCurrent, err, original_);
+    }
+    delete placedContext_;
+}
+
+void ContextCallback(void *userData) {
+    auto *ptr = static_cast<std::atomic<rocblas_handle> *>(userData);
+    if (!ptr) {
+        return;
+    }
+    auto handle = ptr->exchange(nullptr);
+    if (handle != nullptr) {
+        rocblas_status err1;
+        ROCBLAS_ERROR_FUNC(rocblas_destroy_handle, err1, handle);
+        handle = nullptr;
+    }
+    else {
+        // if the handle is nullptr it means the handle was already destroyed by
+        // the rocblas_handle destructor and we're free to delete the atomic
+        // object.
+        delete ptr;
+    }
+}
+
+rocblas_handle RocblasScopedContextHandler::get_handle(const cl::sycl::queue &queue) {
```

Comment:
Please use sycl:: namespace instead of cl::sycl everywhere.

Reply:
Sure, we'll incorporate the change in the next PR